trieste/acquisition/rule.py

```diff
@@ -2073,6 +2073,191 @@ def update(
+
+    def _compute_global_distances(self) -> None:
+        # Pairwise distances along each axis in the global search space.
+        points = self.global_search_space.points
+        self._global_distances = tf.abs(
+            tf.expand_dims(points, -2) - tf.expand_dims(points, -3)
+        )  # [num_points, num_points, D]
+
+    def _update_neighbors(self) -> None:
+        # Indices of the neighbors within the trust region.
+        neighbors_mask = tf.reduce_all(
+            tf.gather(self._global_distances, self._location_ix) <= self.eps, axis=-1
+        )
+        neighbors_mask = tf.reshape(neighbors_mask, (-1,))
+        self._neighbor_ixs = tf.where(neighbors_mask)
+        self._neighbor_ixs = tf.squeeze(self._neighbor_ixs, axis=-1)
+        # Points within the trust region (including the location point).
+        self._points = tf.gather(self.global_search_space.points, self._neighbor_ixs)
+
+    def initialize(
+        self,
+        models: Optional[Mapping[Tag, ProbabilisticModelType]] = None,
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+    ) -> None:
+        """
+        Initialize the region by sampling a location from the global search space and selecting
+        the neighboring points within the trust region.
+
+        :param models: The model for each tag.
+        :param datasets: The dataset for each tag.
+        """
+
+        datasets = self.select_in_region(datasets)
+        self._init_location()
+        self._step_is_success = False
+        self._init_eps()
+        self._update_neighbors()
+        _, self._y_min = self.get_dataset_min(datasets)
+        self._initialized = True
+
+    def update(
+        self,
+        models: Optional[Mapping[Tag, ProbabilisticModelType]] = None,
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+    ) -> None:
+        """
+        Update this region, including location and neighbors, using the given dataset. If the
+        size of the region is less than the minimum size, re-initialize the region.
+
+        If the new optimum improves over the previous optimum by some threshold (that scales
+        linearly with ``kappa``), the previous acquisition is considered successful.
+
+        If the previous acquisition was successful, the size is increased by a factor
+        ``1 / beta``. Conversely, if it was unsuccessful, the size is reduced by the factor
+        ``beta``.
+
+        :param models: The model for each tag.
+        :param datasets: The dataset for each tag.
+        """
+
+        if not self._initialized or tf.reduce_any(self.eps < self._min_eps):
+            self.initialize(models, datasets)
+            return
+
+        datasets = self.select_in_region(datasets)
+        x_min, y_min = self.get_dataset_min(datasets)
+        self._location_ix = tf.where(
+            tf.reduce_all(self.global_search_space.points == x_min, axis=-1)
+        )
+        self._location_ix = tf.squeeze(self._location_ix, axis=-1)
```

Comment:
this seems like the same code as `location` property?
perhaps a private method that is used both here and in the property?

Reply:
This is the reverse operation to the one in the property. The property is using the location-index to select a point from the global set. Whereas here we are doing a search in the global set to find an index.